### ISSUE TITLE
Add ability to specify another stores

### ISSUE THREAD

**Issue by alexeyvasilyev** [stkent/amplify#160]
Not possible to add another app stores at the moment, e.g. Amazon AppStore.

It is possible to add AmazonStoreRule to show rate message if Amazon App Store installed, however there is no way to intercept clicking on rate me for Amazon App Store (Google Play store is trying to be called). I think some kind of StoreClickListener should be added.

Here is some info to help Amazon integration

    public class AmplifyAmazonStoreRule implements IEnvironmentBasedRule {

        @Override
        public boolean shouldAllowFeedbackPrompt(@NonNull final IEnvironmentCapabilitiesProvider environmentInfoProvider) {
            return isInstalledByAmazonAppStore(Application.this);
        }

        @NonNull
        @Override
        public String getDescription() {
            return "AmazonAppStoreRule";
        }
    }

    public static boolean isInstalledByAmazonAppStore(@NonNull Context context) {
        String installerId = null;
        try {
            installerId = context.getPackageManager().getInstallerPackageName(context.getPackageName());
        } catch (Exception ignored) {
        }
        // For Amazon AppStore sideloaded on Android 6.0 and below, and as well as for Amazon Fire devices, installer package is "com.amazon.venezia".
        // For Amazon AppStore sideloaded on Android 7.0+ installer package is "com.google.android.packageinstaller". In this case
        // we need to get all installed packages and search for "com.amazon.venezia".
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N && "com.google.android.packageinstaller".equals(installerId)) {
            List<ApplicationInfo> packages = context.getPackageManager()
                    .getInstalledApplications(PackageManager.GET_META_DATA);
            for (ApplicationInfo packageInfo : packages) {
                if ("com.amazon.venezia".equals(packageInfo.packageName))
                    return true;
            }
            return false;
        } else {
            return "com.amazon.venezia".equals(installerId);
        }
    }

    public static void startAmazonAppstore(
            @NonNull final Activity activity,
            @NonNull final String packageName) {
        try {
            Intent intent = new Intent(
                    Intent.ACTION_VIEW,
                    Uri.parse("http://www.amazon.com/gp/mas/dl/android?p="
                            + packageName));
            activity.startActivity(intent);
        } catch (Exception ignored) {
        }
    }

**Comment by stkent** [comment]
Yup this is a good feature to add! I am busy through the end of November but will circle back in December and look at implementing this.

**Comment by alexeyvasilyev** [comment]
Let me know if you need any help from my side.

**Comment by stkent** [comment]
Thanks for the reminder - I'll try to work on this at the very end of December, and let you know how I get on!

**Comment by stkent** [comment]
@alexeyvasilyev do you need to be able to fall back to the Google Play Store if the Amazon App Store was not the install source? Thinking through some options now. Thanks!

**Comment by alexeyvasilyev** [comment]
I think yes.

**Comment by stkent** [comment]
Cool, thanks! Working on this change but it's a biggish refactor, aiming for completion early next week :)

**Comment by stkent** [comment]
@alexeyvasilyev please try this version of the library in your app:

https://www.dropbox.com/s/t70cjk711y33r30/amplify-160.aar?dl=0

You should be able to place the `.aar` file in your `app/libs` folder, then add the following configuration in your `build.gradle` file:

```
repositories {
    flatDir {
        dirs 'libs'
    }
}
```

and replace your existing amplify `compile` dependency with

```
dependencies {
    compile(name: 'amplify-160', ext: 'aar')
}
```

The public API has changed quite a bit in this version of the app. Here's a snippet from the updated example app in #163 that should help you make adjustments for testing:

```
final String releasePackageName = "com.github.stkent.testapp";

Amplify.initSharedInstance(this)
        .setPositiveFeedbackCollectors(
                new AmazonAppStoreFeedbackCollector(releasePackageName),
                new GooglePlayStoreFeedbackCollector(releasePackageName))
        .setCriticalFeedbackCollectors(
                new DefaultEmailFeedbackCollector(this, "someone@example.com"))
        .setAlwaysShow(true);
```

Feedback collection will be attempted in the order you specify the collectors. Your environment rule will probably need to be updated to look like this:

```
addEnvironmentBasedRule(EnvironmentBasedRuleComposer.anyOf(new GooglePlayStoreRule(), new AmazonAppStoreRule())
```

Thoughts on functionality or API are appreciated. Cheers.

**Comment by alexeyvasilyev** [comment]
It works. But I just found a better way for starting Amazon AppStore app.
Just use intent
"amzn://apps/android?p=" + packageName
If it cannot be started, then run 
"http://www.amazon.com/gp/mas/dl/android?p=" + packageName

https://developer.amazon.com/public/apis/earn/in-app-purchasing/docs-v2/deep-linking-to-the-amazon-client

**Comment by stkent** [comment]
Ok, I'll update that code, nice find!

**Comment by stkent** [comment]
Still working on this; updating README with the changes. This will be scheduled as 2.0.0.

**Comment by alexeyvasilyev** [comment]
Any news when this feature will be available?

**Comment by stkent** [comment]
I'll try to get it done this coming week(end) :)

**Comment by stkent** [comment]
@alexeyvasilyev this should be available as of version 2.0.0. This was a big release, so please test the new version as soon as you are able to and let me know if you run into any issues. One thing to note is that I backed out changes related to compound environment rules (i.e. the `EnvironmentBasedRuleComposer`) for this version of the app - I'd like to release that in 3.0.0 instead as part of a larger rewrite of the rule logic. You can find the source code I had written for that [here](https://github.com/stkent/amplify/blob/5656a775f3add58c8c7c17566555f9a758b5c858/amplify/src/main/java/com/github/stkent/amplify/tracking/rules/EnvironmentBasedRuleComposer.java) if you need to add it manually in the mean time. It's pretty much what you'd expect, nothing fancy :)

**Comment by alexeyvasilyev** [comment]
Looks like it works.

However I would recommend to use detection if the app was installed by Amazon AppStore instead of just checking if Amazon AppStore installed.

This is how I use at the moment.

```
    private static final String AMAZON_APP_STORE_PACKAGE_NAME       = "com.amazon.venezia";
    private static final String AMAZON_APP_STORE_MSHOP_PACKAGE_NAME = "com.amazon.mShop.android";
    private static final String PACKAGE_INSTALLER_PACKAGE_NAME      = "com.google.android.packageinstaller";


    public static boolean isInstalledByAmazonAppStore(@NonNull Context context) {
        String installerId = null;
        try {
            installerId = context.getPackageManager().getInstallerPackageName(context.getPackageName());
        } catch (Exception ignored) {
        }
        boolean isInstalledByAmazonAppStore;
        // For Amazon AppStore sideloaded on Android 6.0 and below, and as well as for Amazon Fire devices, installer package is "com.amazon.venezia".
        // For Amazon AppStore sideloaded on Android 7.0+ installer package is "com.google.android.packageinstaller". In this case
        // we need to check if "com.amazon.venezia" or "com.amazon.mShop.android" packages exist.
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N && PACKAGE_INSTALLER_PACKAGE_NAME.equals(installerId)) {
            return
                    getPackageInfo(context, AMAZON_APP_STORE_PACKAGE_NAME,       PackageManager.GET_ACTIVITIES) != null ||
                    getPackageInfo(context, AMAZON_APP_STORE_MSHOP_PACKAGE_NAME, PackageManager.GET_ACTIVITIES) != null;
        } else {
            return
                    AMAZON_APP_STORE_PACKAGE_NAME.equals(installerId) ||
                    AMAZON_APP_STORE_MSHOP_PACKAGE_NAME.equals(installerId);
        }
    }
```
